sui_programmability/examples/nft_standard/sources/safe.move

```diff
@@ -0,0 +1,108 @@
+module nft_collection::safe {
+
+    /// A shared object for storing NFT's of type `T`, owned by the holder of a unique `OwnerCap`.
+    /// Permissions to allow others to list NFT's can be granted via TransferCap's and BorrowCap's
+    struct Safe<phantom T> has key {
+        id: UID,
+        /// NFT's in this safe, indexed by their ID's
+        // note: Table and ObjectTable will be added by https://github.com/MystenLabs/sui/issues/4203
+        nfts: ObjectTable<ID, T>,
+        /// For easier naming/retrieval of NFT's. range is a subset of the domain of `nfts`
+        nicknames: Table<String, ID>,
+        /// ID's of NFT's that are currently listed for sale. These can only be borrowed immutably
+        listed: VecSet<ID>,
+        /// ID's of NFT's that are currently borrowed. These cannot be listed for sale while borrowing is active
+        borrowed: VecSet<ID>,
+        /// Valid version for TransferCap's
+        transfer_cap_version: u64,
+        /// Valid version for BorrowCap's
+        borrow_cap_version: u64,
+        /// If true, anyone can place NFT's into this safe.
+        /// If false, deposits require the OwnerCap
+        accept_deposits: bool,
+    }
+
+    /// A unique capability held by the owner of a particular `Safe`.
+    /// The holder can issue and revoke `TransferCap`'s and `BorrowCap`'s.
+    /// Can be used an arbitrary number of times
+    struct OwnerCap has key, store {
+        id: UID,
+        /// The ID of the safe that this capability grants permissions to
+        safe_id: ID,
+        /// Version of this cap.
+        version: u64,
+    }
+
+    /// Gives the holder permission to transfer the nft with id `nft_id` out of
+    /// the safe with id `safe_id`. Can only be used once.
+    struct TransferCap has key, store {
```

Comment:
Would be useful to extend this with `collection_id` so that in trading contracts we don't have to ask for both `Safe<T>` and `TransferCap` to verify that the NFT is of the right collection, we can only ask for `TransferCap` and then `assert` that the ID is what we expect it to be.

Reply:
I might get it wrong here. But if we don't include both, race conditions (when listed on multiple marketplaces) become possible.

Reply:
That's correct for selling. However, for listing and NFT, the `Safe` _might_ not be necessary.